packages/next/src/server/app-render/encryption.ts

```diff
@@ -151,20 +169,44 @@ export async function encryptActionBoundArgs(actionId: string, args: any[]) {
 // Decrypts the action's bound args from the encrypted string.
 export async function decryptActionBoundArgs(
   actionId: string,
-  encrypted: Promise<string>
+  encryptedPromise: Promise<string>
 ) {
+  const encrypted = await encryptedPromise
+  const workUnitStore = workUnitAsyncStorage.getStore()
+
+  let decrypted: string | undefined
+
+  if (workUnitStore) {
+    const cacheSignal =
+      workUnitStore.type === 'prerender' ? workUnitStore.cacheSignal : undefined
+
+    const prerenderResumeDataCache = getPrerenderResumeDataCache(workUnitStore)
+    const renderResumeDataCache = getRenderResumeDataCache(workUnitStore)
+
+    decrypted =
+      prerenderResumeDataCache?.decryptedBoundArgs.get(encrypted) ??
+      renderResumeDataCache?.decryptedBoundArgs.get(encrypted)
+
+    if (!decrypted) {
+      cacheSignal?.beginRead()
+      decrypted = await decodeActionBoundArg(actionId, encrypted)
+      cacheSignal?.endRead()
+      prerenderResumeDataCache?.decryptedBoundArgs.set(encrypted, decrypted)
+    }
+  } else {
+    decrypted = await decodeActionBoundArg(actionId, encrypted)
+  }
+
   const { edgeRscModuleMapping, rscModuleMapping } =
     getClientReferenceManifestForRsc()
 
-  // Decrypt the serialized string with the action id as the salt.
-  const decrypted = await decodeActionBoundArg(actionId, await encrypted)
-
   // Using Flight to deserialize the args from the string.
   const deserialized = await createFromReadableStream(
     new ReadableStream({
       start(controller) {
         controller.enqueue(textEncoder.encode(decrypted))
-        controller.close()
+        // Explicitly don't close the stream here so that hanging promises are
+        // not rejected.
```

Comment:
Shouldn't this be connected to the abort signal so that it can clean up resources?